rabbitvcs/util/strings.py

```diff
@@ -0,0 +1,230 @@
+#
+# This is an extension to the Nautilus file manager to allow better
+# integration with the Subversion source control system.
+#
+# Copyright (C) 2006-2008 by Jason Field <jason@jasonfield.com>
+# Copyright (C) 2007-2008 by Bruce van der Kooij <brucevdkooij@gmail.com>
+# Copyright (C) 2008-2010 by Adam Plumb <adamplumb@gmail.com>
+#
+# RabbitVCS is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# RabbitVCS is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with RabbitVCS;  If not, see <http://www.gnu.org/licenses/>.
+#
+
+"""
+
+Additional strings support.
+
+"""
+
+import sys
+import codecs
+import re
+import six
+import locale
+
+__all__ = ["S", "IDENTITY_ENCODING", "UTF8_ENCODING", "SURROGATEESCAPE"]
+
+unicode_null_string = six.u("")
+non_alpha_num_re = re.compile("[^A-Za-z0-9]+")
+SURROGATE_BASE = 0xDC00
+RE_SURROGATE = re.compile(six.u("[") + six.unichr(SURROGATE_BASE + 0x80) +
+                          six.u("-") + six.unichr(SURROGATE_BASE + 0xFF) +
+                          six.u("]"))
+RE_UTF8 = re.compile("^[Uu][Tt][Ff][ _-]?8$")
+
+#   Codec that maps ord(byte) == ord(unicode_char).
+
+IDENTITY_ENCODING = "latin-1"
+
+
+
+#   An UTF-8 codec that implements surrogates, even in Python 2.
+
+UTF8_ENCODING = "rabbitvcs-utf8"
+
+def utf8_decode(input, errors="strict"):
+    return codecs.utf_8_decode(input, errors, True)
+
+
+def utf8_encode(input, errors="strict"):
+    output = b''
+    pos = 0
+    end = len(input)
+    eh = None
+    while pos < end:
+        n = end
+        m = RE_SURROGATE.search(input, pos)
+        if m:
+            n = m.start()
+        if n > pos:
+            p, m = codecs.utf_8_encode(input[pos:n], errors)
+            output += p
+            pos = n
+        if pos < end:
+            e = UnicodeEncodeError(UTF8_ENCODING,
+                                   input, pos, pos + 1,
+                                   "surrogates not allowed")
+            if not eh:
+                eh = codecs.lookup_error(errors)
+            p, n = eh(e)
+            output += p
+            if n <= pos:
+                n = pos + 1
+            pos = n
+    return (output, len(input))
+
+class utf8_IncrementalEncoder(codecs.IncrementalEncoder):
```

Comment:
Mixing the underscore and camel case like this seems odd.  Any reason why?

Reply:
None! I generally use underscore, and the suffix matches the base class. I can change it too, if you prefer.